tests/serve/test_vllm.py

```diff
@@ -0,0 +1,122 @@
+import re
+
+import pytest
+import torch
+
+from outlines.serve.vllm import RegexLogitsProcessor, _patched_apply_logits_processors
+
+
+class MockTokenizer:
+    vocabulary = {
+        **{chr(i): i for i in range(256)},
+        **{"eos": 256},
+    }
+    special_tokens = {"eos"}
+    eos_token_id = 256
+
+    @property
+    def inverse_vocabulary(self):
+        return {v: k for k, v in self.vocabulary.items()}
+
+    def decode(self, token_ids):
+        return "".join([self.inverse_vocabulary[t] for t in token_ids])
+
+    ####
+    # vLLM tokenizer features
+    ####
+    all_special_tokens = list(special_tokens)
+
+    def convert_tokens_to_string(self, token):
+        return token[0]
+
+    def get_vocab(self):
+        return MockTokenizer.vocabulary
+
+
+class MockModel:
+    tokenizer = MockTokenizer()
+
+
+def sample_from_logits(logits):
+    probs = torch.exp(logits) / torch.sum(torch.exp(logits))
+    return torch.multinomial(probs, 1).item()
+
+
+@pytest.mark.parametrize("forget_logits_processor", [True, False])
+def test_time_regexp(forget_logits_processor):
+    pattern = r"(0?[1-9]|1[0-2]):[0-5]\d\s?(am|pm)?"
+    llm = MockModel()
+    logits_processor = RegexLogitsProcessor(pattern, llm)
+
+    token_ids = []
+    while True:
+        random_scores = -10 + 20 * torch.rand(len(llm.tokenizer.vocabulary))
+
+        # mock "forgetting" the logits processor behavior in
+        # vLLM tensor-parallel world size > 1
+        if forget_logits_processor:
+            logits_processor = RegexLogitsProcessor(pattern, llm)
+
+        logits = logits_processor(
+            input_ids=token_ids,
+            scores=random_scores,
+        )
+        new_token_id = sample_from_logits(logits)
+        if new_token_id == llm.tokenizer.eos_token_id:
+            break
+        token_ids.append(new_token_id)
+
+    assert re.fullmatch(pattern, llm.tokenizer.decode(token_ids)) is not None
+
+
+def test_time_regexp_multiple_samples():
```

Comment:
What is this testing?

Reply:
I observed a lack of stability in sequence order when using beam search with Outlines. This resulted in a new token for one sequence being applied to a different sequence.

This test reproduces that behavior.

I will leave an explanatory doc string.